lldb/source/Core/Debugger.cpp

```diff
@@ -391,8 +392,13 @@ bool Debugger::SetTerminalWidth(uint64_t term_width) {
 
   if (auto handler_sp = m_io_handler_stack.Top())
     handler_sp->TerminalSizeChanged();
-  if (m_statusline)
-    m_statusline->TerminalSizeChanged();
+
+  {
+    // This might get called from a signal handler.
+    std::unique_lock<std::mutex> lock(m_statusline_mutex, std::try_to_lock);
+    if (m_statusline)
```

Comment:
I assume you only intend to call `TerminalSizeChanged` when the lock is acquired? This then needs to be `if (lock && m_statusline)`. Otherwise this has the exact same behavior with or without your change.

But one thing I'm not sure about is, if you see `SIGWINCH` and it tries to change the window size but fails because it can't acquire the lock, when does the terminal size get updated again?  Does it have to wait for another `SIGWINCH` or does this code run again later when the lock is free?

Reply:
I do recognize that this does protect other threads from touching `m_statusline` some of the time, but it seems like this can't protect you from every race?

Reply:
Correct, the purpose of this lock is to protect other threads when this is **not** called from a signal handler. You're right that for this method/thread, this is no better than not locking at all. 

Handling the signal is definitely more important than risking a relatively benign race, which is why this doesn't check if the lock succeeded. In the future, if we have a signal handler thread as @labath suggested in an earlier review, we should call a different method (or pass a flag) so we know when we're called from a signal handler. Today, that's not possible because there's now way to know if we were called for a signal at the SB API level, which that would need to be threaded through.

Reply:
Maybe this would be a good time to introduce the signal catcher thread?

I mean, I think we have all the pieces, and it shouldn't be necessary to build something entirely new. I think you could just create a `MainLoop` instance in the driver, register all the signals with it, and then kick off `Run` in a separate thread.

Reply:
Yeah, it's on my todo list. Do we want to do this in the driver or in LLDB proper? I was imaging the latter, where you'd call an SB API that says "start the signal handler thread".

Reply:
My reasoning for doing it on the lldb_private side was that then I could decide whether to lock or not, but I guess that if we use the signal handling thread, that's not a concern anymore as we will just queue the requests. Let me see what it would take to add this to the driver.

Reply:
https://github.com/llvm/llvm-project/pull/134956

Reply:
The reason I want to do it in the driver is because it's more flexible (and it matches status quo). As a library, you can't really know whether a SIGWINCH was "meant" for you, or if you're embedded in some larger terminal application. So, to do it the right way, it would have to be somehow configurable, and I'd like to avoid getting into the business of providing a general signal catching API.